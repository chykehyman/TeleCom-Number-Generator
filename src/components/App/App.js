import React, { Component } from "react";
import styled from "styled-components";
import Header from "../header/index";
import Section from "../section";

const App = styled.div`
  background-color: #dde;
  min-height: 100vh;
  font-size: calc(10px + 2vmin);
`;

export default class TeleApp extends Component {
  state = {
    inputValue: "",
    phoneNumbers: [],
    error: "",
    sortValue: "asc"   
  };

  handleOnChange = event => {
    const { value } = event.target;
    this.setState({
      inputValue: value
    });
  };
  validateInput = () => {
    const { inputValue } = this.state;
    let errorMessage = "";
    if (inputValue <= 0 || inputValue >= 10000) {
      errorMessage = "enter a number less than 10,000 but greater than 0";
    }
    return errorMessage;
  };
  handleGeneratePhoneNumbers = event => {
    const error = this.validateInput();

    event.preventDefault();
    if (!error) {
      let counter = 0;
      let generatedNumbers = [];
      const generatedNumber = () =>
        "0" + Math.floor(Math.random() * 900000000 + 100000000);
      while (counter < this.state.inputValue) {
        generatedNumbers.push(generatedNumber());
        counter++;
      }
      this.setState({
        phoneNumbers: generatedNumbers,
        inputValue: ""
      });
    }
    this.setState({ error });
  };

  handleSortChange = event => {
    event.preventDefault();
    this.setState(
      {
        sortValue: event.target.value
      },
      () => this.handlePhoneNumberSort()
    );
  };

  handlePhoneNumberSort = () => {
    const { sortValue, phoneNumbers } = this.state;
    let sortedPhoneNumbers;
    if (sortValue === "asc") {
      sortedPhoneNumbers = phoneNumbers.sort((a, b) => a - b);
    } else {
      sortedPhoneNumbers = phoneNumbers.sort((a, b) => b - a);
    }
    this.setState({
      phoneNumbers: sortedPhoneNumbers
    });
  };
  render() {
    const state = this.state;
    return (
      <App>
        <Header />
        <Section
          handleChange={this.handleOnChange}
          generatePhoneNumbers={this.handleGeneratePhoneNumbers}
          handleSortChange={this.handleSortChange}
          {...state}
        />
      </App>
    );
  }
}
