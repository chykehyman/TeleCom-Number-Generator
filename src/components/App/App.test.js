import React from "react";
import ReactDOM from "react-dom";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import App from "./App";

configure({ adapter: new Adapter() });

describe("Test case for <App /> component", () => {
  let shallowWrapper;
  let instance;

  beforeEach(() => {
    shallowWrapper = shallow(<App />);
    instance = shallowWrapper.instance();
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("should render the App component", () => {
    expect(shallowWrapper).toMatchSnapshot();
  });

  it("should set state when handleOnchange is called", () => {
    const event = {
      target: {
        value: 30
      }
    };
    instance.handleOnChange(event);
    const state = shallowWrapper.state();

    expect(state.inputValue).toEqual(30);
  });

  it("should set error when input value is greater than 10000 or less than 1", () => {
    const event = {
      preventDefault: () => {}
    };

    let state;

    shallowWrapper.setState({ inputValue: -2 });
    instance.handleGeneratePhoneNumbers(event);
    state = shallowWrapper.state();

    expect(state.inputValue).toEqual(-2);
    expect(state.error).toEqual(
      "enter a number less than 10,000 but greater than 0"
    );

    shallowWrapper.setState({ inputValue: 30000 });
    instance.handleGeneratePhoneNumbers(event);
    state = shallowWrapper.state();

    expect(state.inputValue).toEqual(30000);
    expect(state.error).toEqual(
      "enter a number less than 10,000 but greater than 0"
    );
  });

  it("should generate phone numbers when input value passes validation", () => {
    const event = {
      preventDefault: () => {}
    };
    shallowWrapper.setState({ inputValue: 5 });
    instance.handleGeneratePhoneNumbers(event);
    const state = shallowWrapper.state();

    expect(state.inputValue).toEqual("");
    expect(state.error).toBe("");
    expect(state.phoneNumbers.length).toBe(5);
  });

  it("should call handlePhoneNumberSort when sort value changes", () => {
    const event = {
      preventDefault: () => {},
      target: {
        value: "desc"
      }
    };
    let state;

    const handlePhoneNumberSortSpy = jest.spyOn(instance, "handlePhoneNumberSort")

    shallowWrapper.setState({ phoneNumbers: ["09974736943", "0786756476"] });
    instance.handleSortChange(event);
    state = shallowWrapper.state();
    expect(state.sortValue).toEqual("desc");
    expect(handlePhoneNumberSortSpy).toHaveBeenCalled()

    instance.handleSortChange({
      ...event,
      target: {
        value: "asc"
      }
    });
    state = shallowWrapper.state();
    expect(state.sortValue).toEqual("asc");
  });
});
