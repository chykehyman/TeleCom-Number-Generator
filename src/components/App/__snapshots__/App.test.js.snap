// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for <App /> component should render the App component 1`] = `
<styled.div>
  <_default />
  <Section
    error=""
    generatePhoneNumbers={[Function]}
    handleChange={[Function]}
    handleSortChange={[Function]}
    inputValue=""
    phoneNumbers={Array []}
    sortValue="asc"
  />
</styled.div>
`;
